### ISSUE TITLE
能否提供一个类似run prediction这样的代码？

### ISSUE THREAD

**Issue by opentld** [fjchange/MIST_VAD#23]
看了之前的问题：https://github.com/fjchange/MIST_VAD/issues/7#issue-925575665
好像不应该改dataset.py吧，应该直接改test.py的load_model_dataset方法，只读取model出来，然后再改eval_UCF方法，从一个mp4文件中直接读取frame，进行预测。
但作为一个伸手党，厚着脸皮问问大神，能不能给个run prediction代码，直接读取一个视频文件，然后预测，然后输出视频结果呢？ ლ(╹◡╹ლ) 

@fjchange

**Comment by opentld** [comment]
这样改了一下test.py，不知道对不对？？  ლ(╹◡╹ლ)

############################################################################################
def load_ckpt(args):
    model=getattr(models,args.MODEL.split('_')[1]+'_SGA_STD')(dropout_rate=args.dropout_rate,expand_k=args.expand_k,
                                  freeze_backbone=False,freeze_blocks=None).cuda().eval()
    opt_level = 'O1'
    amp.init(allow_banned=True)
    optimizer = torch.optim.Adam(model.parameters(), lr=args.lr, weight_decay=args.weight_decay, betas=(0.9, 0.999),
                                 )
    model,optimizer=amp.initialize(model,optimizer,opt_level=opt_level,keep_batchnorm_fp32=None)

    if args.MODEL.split('_')[-1]=='C3D':
        if args.MODEL=='UCF_C3D':
            load_model(model,torch.load(_C.UCF_C3D_MODEL_PATH)['model'])
        else:
            load_model(model,torch.load(_C.SHT_C3D_MODEL_PATH)['model'])
    else:
        if args.MODEL=='UCF_I3D':
            load_model(model,torch.load(_C.UCF_I3D_MODEL_PATH)['model'])
        else:
            load_model(model,torch.load(_C.SHT_I3D_MODEL_PATH)['model'])

    return model

def run_UCF(args, model, video_file):
    if args.vis:
        gradcam=GradCAM(model.Regressor,grad_pp=False)

    vc = cv2.VideoCapture(video_file)
    vid_len = vc.get(cv2.CAP_PROP_FRAME_COUNT)
    
    frames = []

    for i in tqdm(range(int(vid_len))):
        ret, frame = vc.read()
        if ret:
            frames.append(cv2.cvtColor(frame, cv2.COLOR_BGR2RGB))
        else:
            print('Bug Reported!')
            exit(-1)
        
    frames=frames.float().contiguous().view([-1, 3, frames.shape[-3], frames.shape[-2], frames.shape[-1]]).cuda()
    with torch.no_grad():
        scores, feat_maps = model(frames)[:2]

    if args.ten_crop:
        scores = scores.view([-1, 10, 2]).mean(dim=-2)

    for f_idx, (clip, score, feat_map) in enumerate(zip(frames, scores, feat_maps)):
        score = score.float().squeeze()[1].detach().cpu().item()
        if np.isnan(score):
            raise ValueError('NaN in score')

        if args.vis:
            cam_map = gradcam(feat_map)
            cam_path = _C.VIS_DIR+ '/{}.jpg'.format(f_idx)
            cam_clip = visualize_CAM_with_clip(cam_map, clip, (320, 240))
            cv2.imwrite(cam_path, cam_clip)

def run(args, video_file):
    model = load_ckpt(args)

    if args.MODEL.split('_')[0]=='UCF':
        run_UCF(args, model, video_file)
#    else:
#        eval_SHT(model,dataloader)

############################################################################################

if __name__=='__main__':
    args=options.parse_args()
#    test(args)
    video_file = '../video/RoadAccidents004_x264.mp4'
    run(args, video_file)

**Comment by fjchange** [comment]
The dataset.py is to use the h5 file, so you can modify it for using video directly.

**Comment by opentld** [comment]
> The dataset.py is to use the h5 file, so you can modify it for using video directly.

How to rewrite this code, can you give an example? thank you very much!
@fjchange
